src/util.js

```diff
@@ -276,6 +276,45 @@ function copyDescriptors(from, to, filter) {
     }
 }
 
+var _hookIdCounter = 0;
+// This function should not be used for methods that return a value
+// Using this function implies small performance overhead, should be used only
+// for debug methods
+function hookTo(prototypeObject, methodName, extension) {
+    var existingMethodImpl = prototypeObject[methodName];
+    if (typeof existingMethodImpl === "function" &&
+        typeof extension === "function") {
+        if (!existingMethodImpl.extensions) {
+            throw new Error("Trying to override extended method,"+
+                " please use util.js:hookTo() to extend it");
+        }
+        extension._hookExtensionId = ++_hookIdCounter;
+        existingMethodImpl.extensions[extension._hookExtensionId] = extension;
+    } else if (typeof existingMethodImpl === "undefined") {
+        prototypeObject[methodName] = function () {
+            var extensions = prototypeObject[methodName].extensions;
+            for (var key in extensions) {
+                if (extensions.hasOwnProperty(key)) {
+                    extensions[key].apply(this, arguments);
+                }
+            }
+        };
+        // Using object instead of array of functions to allow optimizations
+        prototypeObject[methodName].extensions = {};
+        if (typeof extension === "function") {
+            extension._hookExtensionId = ++_hookIdCounter;
```

Comment:
Maybe @petkaantonov could explain what exactly is optimized out in stack trace in this case?